### PR TITLE
fix: update AWS ACL before uploading screenshots

### PR DESCRIPTION
**Pull request by bapmrl**
## Context

Resolves SM-1272/DT-1272 (#953 and #958): Cannot push changes because screenshot upload returns 403.

## The Solution

1. Always call `ScreenshotsManager.initS3ACL` in `SlicesManager.updateSliceModelScreenshotsInPlace`.

## Checklist before requesting a review
- [x] I hereby declare my code ready for review.
- [ ] If it is a critical feature, I have added tests.
- [x] The CI is successful.
- [ ] If there could backward compatibility issues, it has been discussed and planned.

### CODE REVIEW

File: packages/manager/src/managers/slices/SlicesManager.ts [hunk 1]
```diff
@@ -641,6 +641,9 @@ export class SlicesManager extends BaseManager {
 	): Promise<SharedSlice> {
 		const sliceMachineConfig = await this.project.getSliceMachineConfig();
 
+		// Update the AWS ACL before uploading screenshots as it might have expired
+		await this.screenshots.initS3ACL();
+
```
Comment (lihbr) [inline packages/manager/src/managers/slices/SlicesManager.ts]: I'd maybe move that down the stream right to the `uploadScreenshot` method, this way if there's no new screenshot we don't even need to bother with initing the ACL and other methods relying on `uploadScreenshot` would not fail too.

File: packages/manager/src/managers/screenshots/ScreenshotsManager.ts [hunk 2]
```diff
@@ -118,9 +118,10 @@ export class ScreenshotsManager extends BaseManager {
 	}
 
 	async initS3ACL(): Promise<void> {
-		if (this._s3ACL) {
-			return;
-		}
+		// TODO: we need to find a way to create a new AWS ACL only when necessary (e.g., when it has expired).
+		// if (this._s3ACL) {
+		// 	return;
+		// }
```
Comment (lihbr) [inline packages/manager/src/managers/screenshots/ScreenshotsManager.ts:124]: I assume this would require an endpoint on our S3 API to validate the session, maybe we could add a small cache/debounce here (like a 60s cache) in the meantime to prevent spamming the S3 ACL init endpoint when we're uploading all screenshots at once.

We're doing something similar on the client to dedupe `/api/v2` request (a bit hard to follow at first I admit): https://github.com/prismicio/prismic-client/blob/v7/src/createClient.ts#L1858-L1910

**Review by lihbr — APPROVED** [review]
LGTM, just some suggestions if you want to give it a shot

File: packages/manager/src/managers/slices/SlicesManager.ts (hunk 1 above)
Comment (xrutayisire) [reply packages/manager/src/managers/slices/SlicesManager.ts]: I propose this version:

Screenshot 2023-05-03 at 14 44 18

The goal is to only do it if necessary one time for each "updateSliceModelScreenshotsInPlace" function.
It also prevents us to do the cache management since it will be done only one time for all screenshots.

We will need anyway to improve the check at the TODO mentioned.

File: packages/manager/src/managers/screenshots/ScreenshotsManager.ts (hunk 2 above)
Comment (bapmrl) [reply packages/manager/src/managers/screenshots/ScreenshotsManager.ts:124]: Hello @lihbr,

I think @xrutayisire's solution is the simplest for now.

I'll create an issue in Linear to implement the endpoint to validate the session 🙂

Thank you!

File: packages/manager/src/managers/slices/SlicesManager.ts (hunk 1 above)
Comment (bapmrl) [reply packages/manager/src/managers/slices/SlicesManager.ts]: Thanks!

Comment (lihbr) [reply packages/manager/src/managers/slices/SlicesManager.ts]: Sounds good to me, just a heads up though that this *might* not work because this method is called in parallel (concurrently, not sequentially) inside `pushChanges`: https://github.com/prismicio/slice-machine/blob/bmo/fix-init-s3-acl/packages/manager/src/managers/prismicRepository/PrismicRepositoryManager.ts#L262

If you want to be sure the ACL is inited only once, you could instead move it there, before the line I linked to.

Comment (xrutayisire) [reply packages/manager/src/managers/slices/SlicesManager.ts]: If we put the init function at the top of "pushChanges" function, "updateSliceModelScreenshotsInPlace" it's good because it will be called only once for sure BUT we will not be able to check that we do it only if there is a screenshot to upload :/ Right?

Comment (bapmrl) [reply packages/manager/src/managers/slices/SlicesManager.ts]: Thanks @lihbr, it's better indeed. I've updated the code.

Comment (lihbr) [reply packages/manager/src/managers/slices/SlicesManager.ts]: You're right @xrutayisire, that's a compromise

**Review by xrutayisire — APPROVED** [review]

File: packages/manager/src/managers/slices/SlicesManager.ts (hunk 1 above)
Comment (bapmrl) [reply packages/manager/src/managers/slices/SlicesManager.ts]: At least it works like the init process now 😊 The init process first calls `initS3ACL` and then calls `pushSlice` (in parallel) on each slice.
